File: 250106/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.py
n, m = tuple(map(int, input().split()))
grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

visited = [[0 for _ in range(m)] for _ in range(n)]

def in_range(x, y):
    return 0<=x<n and 0<=y<m

def can_go(x, y):
    if not in_range(x, y):
        return False
    
    if visited[x][y] or grid[x][y] == 0:
        return False
    
    return True

def dfs(x, y):
    dxs, dys = [1, 0], [0, 1]

    for dx, dy in zip(dxs, dys):
        nx, ny = x+dx, y+dy

        if can_go(nx, ny):
            visited[nx][ny] = 1
            dfs(nx, ny)

visited[0][0] = 1
dfs(0, 0)

print(visited[n-1][m-1])

File: 250106/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.cpp
#include <iostream>
#include <vector>
using namespace std;
#define MAX_NM 100
#define DIR_NUM 2

int n, m;
int grid[MAX_NM][MAX_NM];
bool visited[MAX_NM][MAX_NM];

int dx[DIR_NUM] = {1, 0}, dy[DIR_NUM] = {0, 1};

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<m;
}

bool CanGo(int x, int y) {
    if(!InRange(x, y)) return false;
    if(grid[x][y] == 0 || visited[x][y]) return false;

    return true;
}

void DFS(int x, int y) {

    for(int d=0; d<DIR_NUM; d++) {
        int nx = x+dx[d];
        int ny = y+dy[d];

        if(CanGo(nx, ny)) {
            visited[nx][ny] = true;
            DFS(nx, ny);
        }
    }
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n>>m;

    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            cin >> grid[i][j];
        }
    }

    visited[0][0] = true;
    DFS(0, 0);

    if(visited[n-1][m-1]) cout << 1;
    else cout << 0;
    return 0;
}